helix-core/src/main/java/org/apache/helix/messaging/handling/HelixTaskExecutor.java

```diff
@@ -521,29 +523,51 @@ public void finishTask(MessageTask task) {
     }
   }
 
-  private void updateMessageState(List<Message> readMsgs, HelixDataAccessor accessor,
+  private void updateMessageState(Collection<Message> msgsToBeUpdated, HelixDataAccessor accessor,
       String instanceName) {
-    Builder keyBuilder = accessor.keyBuilder();
-    List<String> readMsgPaths = new ArrayList<>();
-    List<DataUpdater<ZNRecord>> updaters = new ArrayList<>();
-    for (Message msg : readMsgs) {
-      readMsgPaths.add(msg.getKey(keyBuilder, instanceName).getPath());
-      _knownMessageIds.add(msg.getId());
-      /**
-       * We use the updater to avoid race condition between writing message to zk as READ state and removing message after ST is done
-       * If there is no message at this path, meaning the message is removed so we do not write the message
-       */
-      updaters.add(currentData -> {
-        if (currentData == null) {
-          LOG.warn(
-              "Message {} targets at {} has already been removed before it is set as READ on instance {}",
-              msg.getId(), msg.getTgtName(), instanceName);
-          return null;
-        }
-        return msg.getRecord();
-      });
+    if (!msgsToBeUpdated.isEmpty()) {
+      Builder keyBuilder = accessor.keyBuilder();
+      List<String> updateMsgPaths = new ArrayList<>();
+      List<DataUpdater<ZNRecord>> updaters = new ArrayList<>();
+      for (Message msg : msgsToBeUpdated) {
+        updateMsgPaths.add(msg.getKey(keyBuilder, instanceName).getPath());
+        /**
+         * We use the updater to avoid race condition between writing message to zk as READ state and removing message after ST is done
+         * If there is no message at this path, meaning the message is removed so we do not write the message
+         */
+        updaters.add(currentData -> {
+          if (currentData == null) {
+            LOG.warn(
+                "Message {} targets at {} has already been removed before it is set as READ on instance {}",
+                msg.getId(), msg.getTgtName(), instanceName);
+            return null;
+          }
+          return msg.getRecord();
+        });
+      }
+      accessor.updateChildren(updateMsgPaths, updaters, AccessOption.PERSISTENT);
+    }
+
+    // Note that only cache the known message Ids after the update to ZK is successfully done.
+    // This is to avoid inconsistent cache.
+
+    // if a message in "NEW" state is updated, then we might need to process it soon.
+    boolean isNewMessageUpdated = false;
```

Comment:
The naming here makes me a bit confusing. This message is updated as NEW because it is cannot be processed right? Maybe call it MessageUpdatedAsNew or something.

Reply:
Good name, let me change it.